### PR TITLE
Allow uploads to be sent in batches to avoid congestion

### PR DESCRIPTION
**Pull request by Zurga**
Should solve #1464 . 

I have tested this with a custom `handle_progress/3` function and it works as expected. 
Uploading over 50 files is no problem as new ones are scheduled when uploads in progress are done.

### CODE REVIEW

**Comment by mcrumm** [comment]
Hi @Zurga – thanks for the PR! I am not sure we are ready to add this feature yet, but if we do add this then I would like to call it `max_concurrency` as it mimics a similar behaviour found elsewhere in the ecosystem, for instance in [`Task.async_stream`](https://hexdocs.pm/elixir/Task.html#async_stream/5). WDYT?

**Comment by Zurga** [comment]
Hi @mcrumm,

`max_concurrency` is fine by me, I will change it to that.

File: assets/js/phoenix_live_view.js [hunk 1]
```diff
@@ -365,11 +365,38 @@ class LiveUploader {
...
+  // Filter out the entries that have already started uploading or are done uploading
+  const entriesToProcess = entries.filter((entry) => !done.includes(entry.ref) && !inProgress.includes(entry.ref))
+  if (entriesToProcess.length === 0) return
+
+  for (let i = 0; i < resp.config.batch_size - inProgress.length; i++) {
+    let entryUploader = uploadEntry(entriesToProcess.shift(), resp, liveSocket)
+    uploaders.push(entryUploader)
+  }
+
+  setTimeout(uploadInBatches, 500, entries, onError, resp, liveSocket, uploaders)
```
[the remark this answers is not in the document]
Comment (Zurga) [reply assets/js/phoenix_live_view.js:399]: What if we schedule the next upload through a callback we provide to `EntryUploader.upload()`? We could split the entries up in the amount of batches set by `max_concurrency` and pass them as an argument to the first the uploaders.

Add extra argument `nextEntries` to `uploadEntry(entry, resp, liveSocket, nextEntries)`
Replace the setTimeout with:
```
entryUploader.upload(entriesToUploadAfter, (entries) =>
  if (entries.length === 0) return
  {uploadEntry(entries.shift(), resp, liveSocket, entries)}
)```

which would result in 
```
class EntryUploader {
  ....
  upload(nextEntries, callback) {
    ...upload things
    callback(nextEntries)
}

[the remark this answers is not in the document]
Comment (Zurga) [reply assets/js/phoenix_live_view.js:399]: Also, would the intensive throttling not also occur with the setTimeout in the EntryUploader.pushChunk function?

[the remark this answers is not in the document]
Comment (Zurga) [reply assets/js/phoenix_live_view.js:399]: Or I could replace setTimeout with a `while` loop. That would be the easiest I guess.

[the remark this answers is not in the document]
Comment (Zurga) [reply assets/js/phoenix_live_view.js:399]: Hmmm, if I set the timeout to more than 4ms, it shouldn't be throttled right? 

"The timer isn't throttled, unless the requested timeout is less than 4ms, and the chain count is 5 or greater, in which case the timeout is set to 4ms. This isn't new; browsers have done this for many years."

[the remark this answers is not in the document]
Comment (mcrumm) [reply assets/js/phoenix_live_view.js:399]: > would the intensive throttling not also occur with the setTimeout in the EntryUploader.pushChunk function?

No. That timer is scheduled in the push reply callback, which means that the next timeout is scheduled within a "network response handler". This is the requirement of newer user agent behaviour (as of Chrome 88) and you can read more about it on this Phoenix PR and the issues linked therein: https://github.com/phoenixframework/phoenix/pull/4309

To your point, we are evaluating whether or not other timers in `phoenix_live_view.js` need to refactored as a result of these user agent changes.

[the remark this answers is not in the document]
Comment (chrismccord) [reply assets/js/phoenix_live_view.js:399]: We also need to consider cancelled entries that attempt to start after they've been canceled

[the remark this answers is not in the document]
Comment (Zurga) [reply assets/js/phoenix_live_view.js:399]: @mcrumm thank you for the explanation.

@chrismccord yes, good point! I will add a check for that before they are scheduled. Is it possible to cancel an entry after the form has been submitted?

[the remark this answers is not in the document]
Comment (Zurga) [reply assets/js/phoenix_live_view.js]: My latest commit hopefully addressed the comments in this thread, I will resolve this conversation.

**Comment by Zurga** [comment]
To handle the cancelled entries, I changed the uploading sequence. The upload channels for the entries are created all at once, but the `upload` function is only called when another entry has finished. This way, entries can be cancelled from the `handle_progress` function after submitting the form.

**Comment by chrismccord** [comment]
Unfortunately, this doesn't solve the DoS vector where all channel processes are allowed to be started immediately and bits pushed on the wire since the client is the one controlling concurrency. We need to communicate back with the LiveView to allow the next batch (or next in batch) to begin.

**Comment by Zurga** [comment]
Fair enough. Would it not be better to have only one UploadChannel per LiveView when speaking about DoS vectors? I can imagine multiple clients wanting to upload files simultaneously also being a DoS vector.

**Comment by chrismccord** [comment]
Multiple clients uploading simultaneously is less of an issue because it takes work to create those clients (new websocket connections). The issue with allowing the client to control concurrency is it allows the DoS vector to increase by an order of magnitude when an attacker can do it "for free", vs expending resources on new connections. Multiplexing uploads on a single channel process is an option. There are slowloris attack considerations in place that the upload channel mitigates, and multiplexing would make that more complicated,  so I think it makes sense to keep 1:1 and have the parent LV sign off on the batches. This can happen via a handshake just like allow_upload, but it definitely increases the complexity.

**Comment by chrismccord** [comment]
So to be clear, we can schedule the batches on the client like you did initially, but we need the server to coordinate whether they can begin. There is a similar check already happening that enforces `max_entries`, but it would need to change to enforce max entries at once, along with max total entries. Does that make sense?

**Comment by chrismccord** [comment]
The check would happen at `LiveView.Channel.register_upload`, which we call from `UploadChannel.join`. You'd use the configuration for max_concurrency sent down to the client to only start that many channels at a given time, and enforce a malicious client on join

**Comment by Zurga** [comment]
Okay, I have the basic concept now I guess But I am still getting some errors that join on a channel is called multiple times, I will continue to work on it.

**Comment by Zurga** [comment]
To be clear I got the join errors only while testing `max_conncurrency`, by deliberately starting more uploads than were allowed.

**Comment by Zurga** [comment]
I will close this, as the source has changed a lot. After re-implementing, I will open a new pull request.

**Pull request closed without merging**